src/Features/CSharp/Portable/AssignOutParameters/AssignOutParametersCodeFixProvider.cs

```diff
@@ -0,0 +1,189 @@
+        {
+            for (var current = node; current != null; current = current.Parent)
+            {
+                var parameterList = CSharpSyntaxGenerator.GetParameterList(current);
+                if (parameterList != null)
+                {
+                    return current;
+                }
+            }
+
+            return null;
+        }
+
+        protected override async Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor, CancellationToken cancellationToken)
+        {
+            var root = editor.OriginalRoot;
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var containersAndLocations =
+                diagnostics.SelectAsArray(d => GetContainer(root, d.Location.SourceSpan))
+                           .WhereAsArray(t => t.container != null);
+
+            foreach (var group in containersAndLocations.GroupBy(t => t.container))
+            {
+                var container = group.Key;
+
+                var parameterList = CSharpSyntaxGenerator.GetParameterList(container);
+                var outParameters =
+                    parameterList.Parameters.Select(p => semanticModel.GetDeclaredSymbol(p, cancellationToken))
+                                            .Where(p => p?.RefKind == RefKind.Out)
+                                            .ToImmutableArray();
+
+                var distinctExprsOrStatements = group.Select(t => t.exprOrStatement).Distinct();
+                foreach (var exprOrStatement in distinctExprsOrStatements)
+                {
+                    var dataFlow = semanticModel.AnalyzeDataFlow(exprOrStatement);
+                    var definitelAssignedOnExit = dataFlow.DefinitelyAssignedOnEntry.AddRange(dataFlow.AlwaysAssigned);
+
+                    var unassignedParameters = outParameters.WhereAsArray(p => !definitelAssignedOnExit.Contains(p));
+                    if (unassignedParameters.Length > 0)
+                    {
+                        AssignOutParameters(editor, exprOrStatement, unassignedParameters);
+                    }
+                }
+            }
+        }
+
+        private void AssignOutParameters(
+            SyntaxEditor editor, SyntaxNode exprOrStatement, ImmutableArray<IParameterSymbol> unassignedParameters)
+        {
+            var generator = editor.Generator;
+            var statements = ArrayBuilder<SyntaxNode>.GetInstance();
+
+            foreach (var parameter in unassignedParameters)
+            {
+                statements.Add(generator.ExpressionStatement(generator.AssignmentStatement(
+                    generator.IdentifierName(parameter.Name),
+                    ExpressionGenerator.GenerateExpression(parameter.Type, value: null, canUseFieldReference: false))));
+            }
+
+            var parent = exprOrStatement.Parent;
+            if (parent.IsEmbeddedStatementOwner())
```

Comment:
Would it be possible to extract out the block of code here (line 144-167) to a service/extension method somewhere? I have needed to do a similar syntax transformation couple of times in different fixers/refactorings, and it is unfortunate that understanding of these very special parent node kinds, wrapping in a block, etc. needs to be handled by all clients. Does not need to be done as part of this PR though...

Reply:
Yes.  100%.  in fact, we may have it already in the generator.  If not, i'll move there.

Reply:
Ok.  I did not do this now.  But i can try to in an followup PR.